### PR TITLE
fix umm memory managment

### PR DESCRIPTION
**Pull request by Links2004**
wrap all SDK memory functions in all libs, to get umm to work
fix umm_malloc.c:1708: undefined reference to `putchar'

### CODE REVIEW

**Comment by Links2004** [comment]
will fix #1629 and problems from #1586

**Comment by igrr** [comment]
pvPortMalloc and friends are already defined here: https://github.com/esp8266/Arduino/blob/master/cores/esp8266/heap.c, and corresponding functions from libmain.a are marked as weak, so is additional wrapping required?

**Comment by igrr** [comment]
Regarding printf: sorry for breaking the build, but let's fix the real issue (printf not compiling) instead of replacing it with ets_printf.

**Comment by me-no-dev** [comment]
I use umm without any wrapping at all. Have just the Port stuff weakened and overwritten. All has been working nice since some time before you even introduced it in the main repo. Have additionally added alignment check and put all in RAM and that's about it. :) 2 bytes poison and integrity check are on

**Comment by Links2004** [comment]
yes the wrapping is needed some of my projects only boot when the wrapping is here.
and all mem_* stuff is normally in the rom of the chip, for this we need wrapping too.

**Comment by Links2004** [comment]
to get printf work it is normally not more then this:
```cpp
int putchar_cpp(int c) {
    if(Serial.write(c)){
        return c;
    }
    return -1;
}

extern "C" {
int putchar(int c) {
    return putchar_cpp(int c);
}
}
```

but i think debug an error messages are better in os_printf since it easy to disable the output when serial is needed for something else, and we have already many debug out at os_printf.
and it is better when the user can select where printf is printing.

**Comment by igrr** [comment]
Okay, figured that out. The issue has actually been triggered by 737f6c2, in a very interesting manner.

TL;DR: Depending on the libraries used by the sketch, weakening pvPortMalloc/vPortFree may be not sufficient.

Here is a simple sketch which will fail to run (stripped down version of the one provided by @Links2004):

```c++
#include <Arduino.h>
#include <ESP8266WiFi.h>

void setup() {
    Serial.begin(115200);
    Serial.setDebugOutput(true);
    Serial.println();
    Serial.println();
    Serial.println();
    for(uint8_t t = 4; t > 0; t--) {
        Serial.printf("[SETUP] BOOT WAIT %d...\n", t);
        delay(1000);
    }
    Serial.printf("[SETUP] HEAP: %d\n", ESP.getFreeHeap());
    WiFi.mode(WIFI_AP);
    WiFi.softAP("ESP_SLDT", "yesSecureHere");
}

void loop() { }
```

At the same time, more complex sketches (e.g. FSBrowser sample) seem to work fine.
The reason this happens has to do with linker search order. If none of the object files before `libmain.a` reference `pvPortMalloc` symbol, it will not be pulled from `arduino.ar` (archive file containing the core library). Instead, linker will loop inside libmain.a until `pvPortMalloc`/`vPortFree` are found (that's the rule for dealing with archive files: loop until no additional symbols are resolved). If the sketch uses a library which calls `pvPortMalloc` (usually this happens via `os_malloc` and `os_free` defines), for instance ESP8266mDNS, linker will process it first, and will resolve the reference to `pvPortMalloc` via arduino.ar, pulling the "right" definition. From that moment on, the right definition will be used elsewhere.

I'm going to remove mem_manager.o from libmain.a completely, so that the "right" symbols are always used.
